### PR TITLE
add gomaxprocs limit, return node fit error and pod QoS in advance

### PR DESCRIPTION
**Pull request by fanhaouu**
This PR aims to address the following three issues：
1. The default value of GOMAXPROCS is the number of CPU cores of the machine. After GOMAXPROCS is higher than the number of truly usable cores, the Go scheduler will keep switching OS threads, which will affect the performance of descheduler program;
2. There are many filter policies in the current node fit, similar to the filter plugin in the scheduler. The correct logic should be: if any plugin does not meet the conditions, it should be terminated in advance. The current check logic is incomprehensible, and it actually performs all filter check logic, which is not only time-consuming, but also very unnecessary.(https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/node/node.go#L107)
3. If the pod's Qos has a value, return the QoS value early. For pods evicted within the descheduler, they are typically already scheduled onto nodes, and in most cases, the QoS value is already present on the pod's status. Therefore, it is unnecessary to start the determination process from scratch.

### CODE REVIEW

**Comment by fanhaouu** [comment]
@a7i ,master, can you help me review this pr?

**Comment by a7i** [comment]
Hi @fanhaouu great contribution. Going to copy some of the maintainers for feedback as well:
/cc @jklaw90 @ingvagabund  @damemi

My feedback:
- _regarding point 1_: For `GOMAXPROCS`, as far as I know, we don't run any goroutines in Descheduler. How does this help?

- _regarding point 2_: The idea is to present all NodeFit predicates. The predicate check is not in any particular order, so returning the first one may not present the whole picture to the cluster operator. In a cluster of 35k pods / 15k deployments, a Descheduler run takes a few (single digit) seconds. Given that the shorter run interval can be a minute, I'm not convinced this optimization is worth it. What do you think?

- _regarding point 3_: I like it, that's a great change!

- overall: if you could split this into 3 PRs, I think it would make it easier to provide feedback and request changes.

**Comment by fanhaouu** [comment]
> Hi @fanhaouu great contribution. Going to copy some of the maintainers for feedback as well: /cc @jklaw90 @ingvagabund @damemi
> 
> My feedback:
> 
> * _regarding point 1_: For `GOMAXPROCS`, as far as I know, we don't run any goroutines in Descheduler. How does this help?
> * _regarding point 2_: The idea is to present all NodeFit predicates. The predicate check is not in any particular order, so returning the first one may not present the whole picture to the cluster operator. In a cluster of 35k pods / 15k deployments, a Descheduler run takes a few (single digit) seconds. Given that the shortest frequency can be a minute, I'm not convinced this optimization is worth it. What do you think?
> * _regarding point 3_: I like it, that's a great change!
> * overall: if you could split this into 3 PRs, I think it would make it easier to provide feedback and request changes.



> Hi @fanhaouu great contribution. Going to copy some of the maintainers for feedback as well: /cc @jklaw90 @ingvagabund @damemi
> 
> My feedback:
> 
> * _regarding point 1_: For `GOMAXPROCS`, as far as I know, we don't run any goroutines in Descheduler. How does this help?
> * _regarding point 2_: The idea is to present all NodeFit predicates. The predicate check is not in any particular order, so returning the first one may not present the whole picture to the cluster operator. In a cluster of 35k pods / 15k deployments, a Descheduler run takes a few (single digit) seconds. Given that the shortest frequency can be a minute, I'm not convinced this optimization is worth it. What do you think?
> * _regarding point 3_: I like it, that's a great change!
> * overall: if you could split this into 3 PRs, I think it would make it easier to provide feedback and request changes.

master, thank you for your reply.

point1. At present, descheduler has some goroutine, but the number is not very large, so gomaxproc can be added or not, but the current go runtime cannot recognize the container environment, I think it is better to add it. For example, jvm runtime has long supported container environments; (As descheduler in our company has a lot of goroutine, the performance can be improved significantly after adding this restriction, so I didn't remove this submission from the pr)

point2. Because the default loop time is 5min, this time can be adjusted. The larger the cluster, the longer the nodeFit time, so the full loop time can easily exceed 5min. What if the user sets the loop time to 1 minute? Comparing one by one I think is very, very unnecessary, too time-consuming, we should be the same as the filter policy in the scheduler.

In short, if we can make the program better, faster, without affecting any functionality, I think it makes a lot of sense.

We are also developing a descheduler cache mechanism internally. Currently, descheduler pulls weight data for filtering at policy runtime, which can actually cache a good part of the data in advance, just like the cache in the scheduler. If our company is stable online, I will also contribute to our descheduler community. We can review it together then.

**Comment by ingvagabund** [comment]
As Amir mentioned would you please break the PR into three separate PRs? Some of the suggested changes deserves a dedicated discussion. Wrt. NodeFit I am in the process of composing a KEP: https://github.com/kubernetes-sigs/descheduler/issues/1421. This sounds like a good use case to include in the proposal.

**Comment by fanhaouu** [comment]
hi, masters, i have split this into 3 PRs, looking forward to your feedback:

1. https://github.com/kubernetes-sigs/descheduler/pull/1434
2. https://github.com/kubernetes-sigs/descheduler/pull/1435
3. https://github.com/kubernetes-sigs/descheduler/pull/1436

/cc @a7i @jklaw90 @ingvagabund @damemi

**Comment by a7i** [comment]
/close

it's already split into 3 PRs
